packages/vite/src/node/build.ts

```diff
@@ -1158,14 +1162,20 @@ export function injectEnvironmentToHooks(
   return clone
 }
 
-function wrapEnvironmentResolveId(
+function wrapEnvironmentAndFilterResolveId(
   environment: BuildEnvironment,
   hook?: Plugin['resolveId'],
 ): Plugin['resolveId'] {
   if (!hook) return
 
+  const rawIdFilter = typeof hook === 'object' ? hook.filter?.id : undefined
+  const idFilter = rawIdFilter ? createIdFilter(rawIdFilter) : undefined
+
   const fn = getHookHandler(hook)
   const handler: Plugin['resolveId'] = function (id, importer, options) {
+    if (idFilter && !idFilter(id)) {
+      return
+    }
```

Comment:
nice that you can do this without performance penalty over the current environment injection